src/main/java/htsjdk/samtools/cram/structure/SubstitutionMatrix.java

```diff
@@ -20,160 +20,251 @@
 import java.util.Arrays;
 import java.util.Comparator;
 
+/**
+ * Substitution matrix, used to represent base substitutions for reference-based CRAM
+ * compression.
+ *
+ * The matrix is stored internally in two forms; the packed/encoded form used for serialization,
+ * and an expanded in-memory form used for fast bi-directional interconversion between bases and
+ * substitution codes during read and write.
+ *
+ * This implementation substitutes both upper and lower case versions of a give base with the
+ * same (upper case) substitute base.
+ */
 public class SubstitutionMatrix {
-    public static final byte[] BASES = new byte[]{'A', 'C', 'G', 'T', 'N'};
-    private static final byte[] BASES_LC = new byte[]{'a', 'c', 'g', 't', 'n'};
-    private static final byte[] ORDER;
+    // List of bases subject to substitution. The upper and lower case arrays must have the bases
+    // in the same order so that the index for a given base is the same for upper and lower case
+    public static final byte[] SUBSTITUTION_BASES_UPPER = new byte[]{'A', 'C', 'G', 'T', 'N'};
+    private static final byte[] SUBSTITUTION_BASES_LOWER = new byte[]{'a', 'c', 'g', 't', 'n'};
+    private static int BASES_SIZE = SUBSTITUTION_BASES_UPPER.length;
+
+    // Since bases are represented as bytes, there are theoretically 256 possible symbols in the
+    // symbol space, though in reality we only care about 10 of these (5 upper and 5 lower case
+    // reference bases).
+    private static int SYMBOL_SPACE_SIZE = 256;
 
+    // number of possible substitution codes per base
+    private static int CODES_PER_BASE = BASES_SIZE - 1;
+
+    // The order in which substitution codes are encoded in the serialized matrix, as prescribed
+    // by the CRAM spec:
+    private static final byte[] CODE_ORDER;
     static {
-        ORDER = new byte[255];
-        Arrays.fill(ORDER, (byte) -1);
-        ORDER['A'] = 0;
-        ORDER['C'] = 1;
-        ORDER['G'] = 2;
-        ORDER['T'] = 3;
-        ORDER['N'] = 4;
+        CODE_ORDER = new byte[SYMBOL_SPACE_SIZE];
```

Comment:
This whole thing is very fragile if we ever change the bases (which we probably don't...) it could be rewritten to actually use the values  and indexes from the `SUBSTITUTION_BASES_UPPER` automatically.  That might be less clear though...

Reply:
Yes, I plan to do exactly that in the next PR .